lib/auth.js

```diff
@@ -0,0 +1,189 @@
+/*!
+ * Tozny e3db
+ *
+ * LICENSE
+ *
+ * Tozny dual licenses this product. For commercial use, please contact
+ * info@tozny.com. For non-commercial use, the contents of this file are
+ * subject to the TOZNY NON-COMMERCIAL LICENSE (the "License") which
+ * permits use of the software only by government agencies, schools,
+ * universities, non-profit organizations or individuals on projects that
+ * do not receive external funding other than government research grants
+ * and contracts.  Any other use requires a commercial license. You may
+ * not use this file except in compliance with the License. You may obtain
+ * a copy of the License at https://tozny.com/legal/non-commercial-license.
+ * Software distributed under the License is distributed on an "AS IS"
+ * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
+ * License for the specific language governing rights and limitations under
+ * the License. Portions of the software are Copyright (c) TOZNY LLC, 2017.
+ * All rights reserved.
+ *
+ * @copyright Copyright (c) 2017 Tozny, LLC (https://tozny.com)
+ */
+
+/* global fetch */
+
+import 'isomorphic-fetch'
+
+import { default as uuid } from 'uuid/v4'
+
+const supportedTSV1Algorithms = ['TSV1-ED25519-BLAKE2B']
+
+/**
+ * Fallback polyfill to allow for HTTP Basic authentication from either Node
+ * or browser-based JavaScript.
+ *
+ * @param {string} str String to encode as Base64
+ */
+function btoa(str) {
+  let buf = Buffer.from(str, 'utf8')
+  return buf.toString('base64')
+}
+
+function sortQueryParams(queryString) {
+  // Parse query string into field=values
+  var query = queryString.split('&')
+  for (var i = 0; i < query.length; i++) {
+    // Separate fields from values
+    let splitQuery = query[i].split('=')
+    query[i] = splitQuery
+  }
+  // Sort alphabetically
+  query.sort(function(a, b) {
+    return a[0] < b[0] ? -1 : 1
+  })
+  // Reconstruct query
+  for (var j = 0; j < query.length; j++) {
+    query[j] = query[j].join('=')
+  }
+  return query.join('&')
+}
+
+class Auth {
```

Comment:
This is a really generic name. I'd like to see the helper named something a little more descriptive of what it is providing.

Reply:
Would `AuthenticationHelper` be more descriptive to you? One thing about the structure of these SDKs that I struggled with is the PreClient and how we're going to do static auth methods (anonymous notes) within this implementation; the way I have it written you'd have to initialize a partial client within that method.

Now that I've stepped back a little bit, I realize I can have a PreAuth object that the libSodium sdk will export in addition to the `PreClient` object. This would separate the ties between the client and auth particularly for signature methods. Its a little weird that this helper can also do token auth (very client centric), but maybe a worthwhile tradeoff?